src/common/variant.h

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2022, The Monero Project
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// Variant wrapper class.
+
+#pragma once
+
+//local headers
+
+//third party headers
+#include <boost/blank.hpp>
+#include <boost/none_t.hpp>
+#include <boost/variant/apply_visitor.hpp>
+#include <boost/variant/get.hpp>
+#include <boost/variant/static_visitor.hpp>
+#include <boost/variant/variant.hpp>
+
+//standard headers
+#include <stdexcept>
+#include <type_traits>
+
+//forward declarations
+
+
+namespace tools
+{
+
+////
+// variant: convenience wrapper around boost::variant with a cleaner interface
+// - the value may be assigned to but is otherwise read-only
+// - the variant is 'optional' - an empty variant will evaluate to 'false' and an initialized variant will be 'true'
+///
+template <typename ResultT>
+struct variant_static_visitor : public boost::static_visitor<ResultT>
+{
+    /// provide visitation for empty variants
+    /// - add this to your visitor with: using variant_static_visitor::operator();
+    ResultT operator()(const boost::blank) const
+    {
+        throw std::runtime_error("variant: tried to visit an empty variant.");
+        static const ResultT dummy{};
+        return dummy;
+    }
+};
+
+template <typename... Types>
+class variant final
+{
+    using VType = boost::variant<boost::blank, Types...>;
```

Comment:
`boost::blank` is designed specifically for this use-case afaik. Plus, a variant isn't a pointer class so I think `nullptr` is a little inconsistent.

Reply:
I would continue to use `boost::blank`. An overload with `nullptr_t` (not needed now), could definitely be useful in generic code like in my serialization library. There are situations where the concepts between pointer, optional, and variant are nearly identical - assign this thing a "dead" value - but with three different techniques for assignment, requires 3 independent functions until C++17 `constexpr if` (which also confuses typical C++ devs).

Anyway, this is why I always thought `std::nullptr_t`/`nullptr` has always seemed just as appropriate though. `std::optional` has `std::nullopt_t` and `nullopt`, and the only difference is slightly more accurate type-checking but little else. Yes, `std::optional` is not _literally_ a pointer, but the intended objectives are identical and why get so fancy with unique types (because it makes writing template/generic functions _more_ complicated).